rasa/nlu/featurizers/sparse_featurizer/regex_featurizer.py

```diff
@@ -42,31 +44,133 @@ def required_components(cls) -> List[Type[Component]]:
+            pattern_vocabulary_stats: Statistics about number of pattern slots filled and total number available.
+            finetune_mode: Load component in finetune mode.
+        """
         super().__init__(component_config)
 
         self.known_patterns = known_patterns if known_patterns else []
         self.case_sensitive = self.component_config["case_sensitive"]
+        self.number_additional_patterns = self.component_config[
+            "number_additional_patterns"
+        ]
+        self.finetune_mode = finetune_mode
+        self.pattern_vocabulary_stats = pattern_vocabulary_stats
+
+        if self.finetune_mode and not self.pattern_vocabulary_stats:
+            # If the featurizer is instantiated in finetune mode,
+            # the vocabulary stats for it should be known.
+            raise RasaException(
+                f"{self.__class__.__name__} was instantiated with"
+                f" `finetune_mode=True` but `pattern_vocabulary_stats`"
+                f" was left to `None`. This is invalid since the featurizer"
+                f" needs vocabulary statistics to featurize in finetune mode."
+            )
+
+    @lazy_property
+    def vocabulary_stats(self) -> Dict[Text, int]:
+        """Compute total vocabulary size and how much of it is consumed.
+
+        Returns:
+            Computed vocabulary size and number of filled vocabulary slots.
+        """
+        if not self.finetune_mode:
+            max_number_patterns = (
+                len(self.known_patterns) + self._get_num_additional_slots()
+            )
+            return {
+                "pattern_slots_filled": len(self.known_patterns),
+                "max_number_patterns": max_number_patterns,
+            }
+        else:
+            self.pattern_vocabulary_stats["pattern_slots_filled"] = len(
+                self.known_patterns
+            )
+            return self.pattern_vocabulary_stats
+
+    def _merge_new_patterns(self, new_patterns: List[Dict[Text, Text]]):
+        """Update already known patterns with new patterns extracted from data.
+
+        Args:
+            new_patterns: Patterns extracted from training data and to be merged with known patterns.
+        """
+        max_number_patterns = self.pattern_vocabulary_stats["max_number_patterns"]
+        pattern_name_index_map = {
+            pattern["name"]: index for index, pattern in enumerate(self.known_patterns)
+        }
+        patterns_dropped = False
+
+        for extra_pattern in new_patterns:
+            new_pattern_name = extra_pattern["name"]
+
+            # Some patterns may have just new examples added
+            # to them. These do not count as additional pattern.
+            if new_pattern_name in pattern_name_index_map:
+                self.known_patterns[pattern_name_index_map[new_pattern_name]][
+                    "pattern"
+                ] = extra_pattern["pattern"]
```

Comment:
Does `extra_pattern["pattern"]` contain all the patterns? Or should it rather be combined with the previous known patterns?

Reply:
`extra_pattern["pattern"]` is one of the patterns present in the training data. Particularly at this stage of the code, this is a pattern which was present with the same `name` before in a previous training run as well. So here we just take the new regex(`extra_pattern["pattern"]`) of the pattern.

Reply:
But don't we overwrite it instead of adding it? Ah no, we just have one pattern for every pattern name 💡 Then it is fine 👍